whois-commons/src/main/java/net/ripe/db/whois/common/rpki/RpkiService.java

```diff
@@ -20,16 +21,27 @@ public class RpkiService {
 
     private static final Logger LOGGER = LoggerFactory.getLogger(RpkiService.class);
 
-    private final NestedIntervalMap<Ipv4Resource, Set<Roa>> ipv4Tree = new NestedIntervalMap<>();
-    private final NestedIntervalMap<Ipv6Resource, Set<Roa>> ipv6Tree = new NestedIntervalMap<>();
+    private NestedIntervalMap<Ipv4Resource, Set<Roa>> ipv4Tree;
+    private NestedIntervalMap<Ipv6Resource, Set<Roa>> ipv6Tree;
+
+    private final RpkiDataProvider rpkiDataProvider;
 
     public RpkiService(final RpkiDataProvider rpkiDataProvider) {
+        this.rpkiDataProvider = rpkiDataProvider;
+        loadRoas();
+    }
+
+    @Scheduled(cron = "0 */15 * * * *")
+    private void loadRoas() {
         final List<Roa> loadedRoas = rpkiDataProvider.loadRoas();
         if (loadedRoas != null && !loadedRoas.isEmpty()){
             final List<Roa> roas = loadedRoas.stream()
                     .filter(roa -> roa.getTrustAnchor() != TrustAnchor.UNSUPPORTED)
                     .collect(Collectors.toList());
 
+            ipv4Tree = new NestedIntervalMap<>();
```

Comment:
This is still clearing the map until it is updated, which will cause problems for asynchronous queries. 
Could you split the update by tree, e.g.
```
this.ipv4Tree = buildTree(...);
this.ipv6Tree = buildTree(...);
```
and within buildTree create a new NestedIntervalMap and return it? Then the map instance is replaced in one operation.

Reply:
talk via Mattermost. I will try to avoid looping twice same list.
So we will go with overlapping those variables but in the method that just loops once